### ISSUE TITLE
Default implementation for ILogger

### ISSUE THREAD

**Issue by dotnetjunkie** [serilog/serilog#861]
Serilog's core `ILogger` interface contains 79 members. This makes it extremely cumbersome to create custom implementation, since all 79 members need to be implemented. This design is very similar to log4net that has an `ILog` interface with the same kind of structure.

To work around the problem that the wide interface causes, log4net contains a default implementation of `ILog` called `LogImpl` . `LogImpl` allows overriding individual members and forwards calls to a wrapped logger. Serilog however seems to lack such default implementation, which forces the user to create such implementation itself. This is cumbersome, error prone and far from a good way to spend your time as a developer.

There are a wide range of reasons to create custom implementations, for instance to create a generic implementation of `ILogger` that can be registered in your DI library. Take for instance [this generic log4net `ILog` implementation](https://stackoverflow.com/questions/25112251/log4net-with-di-simple-injector) that can be registered contextually in Simple Injector.

Other reasons to create such wrapper is to decorate instances, or to override individual members of the implementation.

I suggest to create an `LogImpl` equivalent for Serilog. This `LogImpl` could by default forward calls to `Log.Logger`, while allowing a user to override this behavior.

Example:

``` c#
/// <summary>
/// Implementation of logger that forwards the calls to the wrapped <see cref="Logger"/>.
/// </summary>
public class LogImpl : ILogger
{
    /// <summary>Returns the logger to forward to.</summary>
    public virtual ILogger Logger => Log.Logger;

    /// <summary>Create a logger that enriches log events via the provided enrichers.</summary>
    /// <param name="enricher">Enricher that applies in the context.</param>
    /// <returns>A logger that will enrich log events as specified.</returns>
    public virtual ILogger ForContext(ILogEventEnricher enricher)
    {
        return this.Logger.ForContext(enricher);
    }

    // etc.
}
```

This would allow a user to create the following implementation:

``` c#
public sealed class LogImpl<T> : LogImpl
{
    private static readonly ILogger logger = Log.Logger.ForContext<T>();
    public override ILogger Logger { get { return logger; } }
}
```

You might even want to include this `LogImpl<T>` OOTB as well.

**Comment by nblumhardt** [comment]
Thanks for raising the ticket! Definitely in favour of simplifying this use case. (Changed the title to avoid confusing it with a bug.)

Rather than focus on delegation as the primary use case, perhaps instead the base class could provide concrete implementations of the many overloads, but funnel all calls into one of a few primary versions of: `IsEnabled()`, `Write()`, `Emit()`, `ForContext()`, `Bind*()` and so-on?

The delegation scenario above would then look something like:

```csharp
public sealed class LogImpl<T> : LogImp
{
    private static readonly ILogger logger = Log.Logger.ForContext<T>();

    public override void Write(LogEvent logEvent)
    {
        this.logger.Write(logEvent);
    }
}
```

There may be a few details to work out - we'd want invariants to hold by default so that, for example, if I override `IsEnabled()` whatever custom logic I use is applied by the default implementations of `Write()` and so-on.

Perhaps if delegation is a common case there could be a built-in `DelegatingLogger` that inherits from `LogImpl`?

(`LogImpl` might not be the optimal name good enough for a working title :-) )

**Comment by dotnetjunkie** [comment]
I agree with your points. Such default base class could serve several use cases; in my proposal it only serves one use case. And obviously `LogImpl` is a really - really bad name.

**Comment by nblumhardt** [comment]
Any further thoughts on this one Steven? Cheers!

**Comment by dotnetjunkie** [comment]
I'm sorry Nicholas, I'm afraid I've been trying to do more than I could chew. I've been quite busy with the whole ASP.NET Core Conforming Container experience, and of course my own projects. I'm unsure whether I will be able to create a PR for this.

**Comment by nblumhardt** [comment]
Thanks for the update - I know that feeling! :-)

I'll close this for now, but if you or anyone wants to pick it up in the future we can give it a go then. All the best!

**Comment by emumanu** [comment]
There you have it @dotnetjunkie 

```cs
public class LogImpl : ILogger
{
    private ILogger _logger;

    /// <inheritdoc />
    public virtual ILogger ForContext(ILogEventEnricher enricher)
    {
        return _logger.ForContext(enricher);
    }

    /// <inheritdoc />
    public virtual ILogger ForContext(IEnumerable<ILogEventEnricher> enrichers)
    {
        return _logger.ForContext(enrichers);
    }

    /// <inheritdoc />
    public virtual ILogger ForContext(string propertyName, object value, bool destructureObjects = false)
    {
        return _logger.ForContext(propertyName, value, destructureObjects);
    }

    /// <inheritdoc />
    public virtual ILogger ForContext<TSource>()
    {
        return _logger.ForContext<TSource>();
    }

    /// <inheritdoc />
    public virtual ILogger ForContext(Type source)
    {
        return _logger.ForContext(source);
    }

    /// <inheritdoc />
    public virtual void Write(LogEvent logEvent)
    {
        _logger.Write(logEvent);
    }

    /// <inheritdoc />
    public virtual void Write(LogEventLevel level, string messageTemplate)
    {
        _logger.Write(level, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Write<T>(LogEventLevel level, string messageTemplate, T propertyValue)
    {
        _logger.Write(level, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Write<T0, T1>(LogEventLevel level, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Write(level, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Write<T0, T1, T2>(LogEventLevel level, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Write(level, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Write(LogEventLevel level, string messageTemplate, params object[] propertyValues)
    {
        _logger.Write(level, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Write(LogEventLevel level, Exception exception, string messageTemplate)
    {
        _logger.Write(level, exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Write<T>(LogEventLevel level, Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Write(level, exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Write<T0, T1>(LogEventLevel level, Exception exception, string messageTemplate, T0 propertyValue0,
        T1 propertyValue1)
    {
        _logger.Write(level, exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Write<T0, T1, T2>(LogEventLevel level, Exception exception, string messageTemplate, T0 propertyValue0,
        T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Write(level, exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Write(LogEventLevel level, Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Write(level, exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual bool IsEnabled(LogEventLevel level)
    {
        return _logger.IsEnabled(level);
    }

    /// <inheritdoc />
    public virtual void Verbose(string messageTemplate)
    {
        _logger.Verbose(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Verbose<T>(string messageTemplate, T propertyValue)
    {
        _logger.Verbose(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Verbose<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Verbose(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Verbose<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Verbose(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Verbose(string messageTemplate, params object[] propertyValues)
    {
        _logger.Verbose(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Verbose(Exception exception, string messageTemplate)
    {
        _logger.Verbose(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Verbose<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Verbose(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Verbose<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Verbose(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Verbose<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Verbose(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Verbose(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Verbose(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Debug(string messageTemplate)
    {
        _logger.Debug(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Debug<T>(string messageTemplate, T propertyValue)
    {
        _logger.Debug(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Debug<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Debug(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Debug<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Debug(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Debug(string messageTemplate, params object[] propertyValues)
    {
        _logger.Debug(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Debug(Exception exception, string messageTemplate)
    {
        _logger.Debug(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Debug<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Debug(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Debug<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Debug(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Debug<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Debug(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Debug(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Debug(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Information(string messageTemplate)
    {
        _logger.Information(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Information<T>(string messageTemplate, T propertyValue)
    {
        _logger.Information(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Information<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Information(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Information<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Information(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Information(string messageTemplate, params object[] propertyValues)
    {
        _logger.Information(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Information(Exception exception, string messageTemplate)
    {
        _logger.Information(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Information<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Information(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Information<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Information(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Information<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Information(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Information(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Information(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Warning(string messageTemplate)
    {
        _logger.Warning(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Warning<T>(string messageTemplate, T propertyValue)
    {
        _logger.Warning(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Warning<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Warning(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Warning<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Warning(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Warning(string messageTemplate, params object[] propertyValues)
    {
        _logger.Warning(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Warning(Exception exception, string messageTemplate)
    {
        _logger.Warning(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Warning<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Warning(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Warning<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Warning(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Warning<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Warning(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Warning(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Warning(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Error(string messageTemplate)
    {
        _logger.Error(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Error<T>(string messageTemplate, T propertyValue)
    {
        _logger.Error(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Error<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Error(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Error<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Error(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Error(string messageTemplate, params object[] propertyValues)
    {
        _logger.Error(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Error(Exception exception, string messageTemplate)
    {
        _logger.Error(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Error<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Error(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Error<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Error(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Error<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Error(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Error(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Error(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Fatal(string messageTemplate)
    {
        _logger.Fatal(messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Fatal<T>(string messageTemplate, T propertyValue)
    {
        _logger.Fatal(messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Fatal<T0, T1>(string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Fatal(messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Fatal<T0, T1, T2>(string messageTemplate, T0 propertyValue0, T1 propertyValue1, T2 propertyValue2)
    {
        _logger.Fatal(messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Fatal(string messageTemplate, params object[] propertyValues)
    {
        _logger.Fatal(messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual void Fatal(Exception exception, string messageTemplate)
    {
        _logger.Fatal(exception, messageTemplate);
    }

    /// <inheritdoc />
    public virtual void Fatal<T>(Exception exception, string messageTemplate, T propertyValue)
    {
        _logger.Fatal(exception, messageTemplate, propertyValue);
    }

    /// <inheritdoc />
    public virtual void Fatal<T0, T1>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1)
    {
        _logger.Fatal(exception, messageTemplate, propertyValue0, propertyValue1);
    }

    /// <inheritdoc />
    public virtual void Fatal<T0, T1, T2>(Exception exception, string messageTemplate, T0 propertyValue0, T1 propertyValue1,
        T2 propertyValue2)
    {
        _logger.Fatal(exception, messageTemplate, propertyValue0, propertyValue1, propertyValue2);
    }

    /// <inheritdoc />
    public virtual void Fatal(Exception exception, string messageTemplate, params object[] propertyValues)
    {
        _logger.Fatal(exception, messageTemplate, propertyValues);
    }

    /// <inheritdoc />
    public virtual bool BindMessageTemplate(string messageTemplate, object[] propertyValues, out MessageTemplate parsedTemplate,
        out IEnumerable<LogEventProperty> boundProperties)
    {
        return _logger.BindMessageTemplate(messageTemplate, propertyValues, out parsedTemplate, out boundProperties);
    }

    /// <inheritdoc />
    public virtual bool BindProperty(string propertyName, object value, bool destructureObjects, out LogEventProperty property)
    {
        return _logger.BindProperty(propertyName, value, destructureObjects, out property);
    }
}
```

**Comment by netcorefan1** [comment]
Is the [solution](https://github.com/serilog/serilog/issues/861#issuecomment-550182409) posted above still the only way to add some custom functions to serilog? I want to create some overloads, but when I tried to play with ILogger inheritance I failed in the casting process.
It has been a couple of years from this and I'm wondering if in the meantime has been developed some special feature to handle such cases.
